experimental/compactdecimal/src/provider.rs

```diff
@@ -0,0 +1,104 @@
+)]
+#[yoke(prove_covariance_manually)]
+pub struct CompactDecimalPatternDataV1<'data> {
+    /// A map keyed on log10 of the CLDR `type` attribute and the `cldr` count attribute.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub patterns: ZeroMap2d<'data, i8, Count, PatternULE>,
+}
+
+/// A CLDR plural keyword, or the explicit value 1.
+/// See <https://www.unicode.org/reports/tr35/tr35-numbers.html#Language_Plural_Rules>.
+#[zerovec::make_ule(CountULE)]
+#[zerovec::derive(Debug)]
+#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(
+    feature = "datagen", 
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_compactdecimal::provider)
+)]
+#[repr(u8)]
+pub enum Count {
+    /// The CLDR keyword `other`.
+    Zero = 0,
+    /// The CLDR keyword `one`.
+    One = 1,
+    /// The CLDR keyword `two`.
+    Two = 2,
+    /// The CLDR keyword `few`.
+    Few = 3,
+    /// The CLDR keyword `many`.
+    Many = 4,
+    /// The CLDR keyword `other`.
+    Other = 5,
+    /// The explicit 1 case, see <https://www.unicode.org/reports/tr35/tr35-numbers.html#Explicit_0_1_rules>.
+    Explicit1 = 6,
+    // NOTE(egg): No explicit 0, because the compact decimal pattern selection
+    // algorithm does not allow such a thing to arise.
+}
+
+/// A compact decimal pattern, representing some literal text with an optional
+/// placeholder, and the power of 10 expressed by the text.
+#[derive(
+    Debug, Clone, Default, PartialEq, yoke::Yokeable, zerofrom::ZeroFrom, Ord, PartialOrd, Eq,
+)]
+#[zerovec::make_varule(PatternULE)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(
+    feature = "datagen", 
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_compactdecimal::provider),
+    zerovec::derive(Serialize),
+)]
+#[zerovec::derive(Debug)]
+#[cfg_attr(feature = "serde", zerovec::derive(Deserialize))]
+pub struct Pattern<'data> {
+    /// The index in literal_text before which the placeholder is inserted;
+    /// this is 0 for insertion at the beginning, which is most common.
+    /// The value 255 indicates that the pattern does not have a placeholder,
+    /// as in French "mille" for 1000.
+    /// This is derived from the numbers of 0s in the pattern and the associated
+    /// `type` attribute; it is a more convenient representation than the number
+    /// of 0s, because it is often common to multiple types; for instance, the
+    /// following correspond to the same [`Pattern`]:
+    ///   <pattern type="1000000" count="other">0 M</pattern>
+    ///   <pattern type="10000000" count="other">00 M</pattern>
+    pub index: u8,
+    /// The compact decimal exponent, e.g., 6 for "million".
+    /// The value 0 indicates that compact notation is not used; in that case,
+    /// literal text must be empty; this corresponds to the CLDR pattern "0".
+    pub exponent: i8,
```

Comment:
Question: when can `exponent` be 0?

How hard is the requirement that if `exponent` is 0 then `literal_text` is empty? It seems like an unnecessary constraint to put upon ourselves.

Reply:
Exponent is 0 when the CLDR pattern is 0, meaning non-compact notation should be used.
CLDR is incapable of expressing non-compact notation with a pattern; this isn't really an ICU4X constraint, but a CLDR fact.

Reply:
Not that I know of, and so it doesn’t show up in the data because I have an implicit {exponent:0, index:0, literal_text:0} at the beginning of everything.
But it could, and having these 0s is convenient when producing the data before collapsing and removing the implicit stuff. 
It also makes the logic cleaner when looking for the right exponent: look for the highest pattern below the magnitude, default to 0, and you don’t need to worry about whether it is a default-0 or a real 0. The 0 can then be fed directly to CompactDecimal with #2889.